### PR TITLE
IDE-901 Add bookmarks to the ribbon bar

### PR DESCRIPTION
**Pull request by dehilsterlexis**
Also fixed bugs with the bookmarks

### CODE REVIEW

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: eclide/mainfrm.cpp [hunk 1]
```diff
@@ -1057,6 +1075,17 @@ void CMainFrame::InitializeRibbon()
         pPanelLocate->Add(new CMFCRibbonButton(ID_EDIT_REPLACE, _T("Replace\ne"), 5));
         pButtFind->AddSubItem(new CMFCRibbonButton(ID_EDIT_GOTO, _T("Goto\ng"), 30));
 
+        CMFCRibbonPanel* pPanelBookmark = m_CategoryHome->AddPanel(_T("Bookmarks"), m_PanelImages.ExtractIcon(27));
+        CMFCRibbonButton * pButtBook = new CMFCRibbonButton(ID_BOOKMARKS_VIEW, _T("View\nf"), 10, 9);
+        pPanelBookmark->Add(pButtBook);
+        pPanelBookmark->Add(new CMFCRibbonButton(ID_BOOKMARKS_NEXT, _T("Next\no"), 6));
```
Comment (GordonSmith) [inline eclide/mainfrm.cpp]: `o` is not in `Next`

File: eclide/mainfrm.cpp [hunk 2]
```diff
@@ -1057,6 +1075,17 @@ void CMainFrame::InitializeRibbon()
...
+        pPanelBookmark->Add(new CMFCRibbonButton(ID_BOOKMARKS_PREV, _T("Previous\ns"), 7));
+        pPanelBookmark->Add(new CMFCRibbonButton(ID_BOOKMARKS_CLEAR, _T("Clear\nl")));
+        pButtBook->AddSubItem(new CMFCRibbonButton(ID_BOOKMARKS_OPEN, _T("Open\ns")));
+        pButtBook->AddSubItem(new CMFCRibbonButton(ID_BOOKMARKS_SAVE, _T("Save\ns")));
+        pButtBook->AddSubItem(new CMFCRibbonButton(ID_BOOKMARKS_LOAD, _T("Load\nl")));
+        pButtBook->AddSubItem(new CMFCRibbonButton(ID_BOOKMARKS_DELETE, _T("Delete\ns")));
```
Comment (GordonSmith) [inline eclide/mainfrm.cpp]: Your using a lot of `s` for the keyboard shortcuts...

**Comment by richardkchapman** [comment]
Should this be 7.6?

**Comment by GordonSmith** [comment]
Yes - @dehilsterlexis can you re-target to master?

File: eclide/mainfrm.cpp (hunk 1 above)
Comment (dehilsterlexis) [reply eclide/mainfrm.cpp]: Good catch!

File: eclide/mainfrm.cpp (hunk 2 above)
Comment (dehilsterlexis) [reply eclide/mainfrm.cpp]: Ditto...

File: eclide/BookmarksFrame.cpp [hunk 3]
```diff
@@ -37,11 +37,92 @@ bool CBookmarksFrame::OnContext(CMenu &/*menu*/, CPoint &/*pt*/)
...
-    CBookmarksView &bookmarksFrame = (CBookmarksView &)m_view;
-    bookmarksFrame.ParseBookmarksEcl(ecl, user, id, inModule, inAttributeName, attrType);
+    CBookmarksView &bookmarksView = (CBookmarksView &)m_view;
+    bookmarksView.ParseBookmarksEcl(ecl, user, id, inModule, inAttributeName, attrType);
 }
 void CBookmarksFrame::ParseBookmarksEcl(IAttribute *attribute)
 {
-    CBookmarksView &bookmarksFrame = (CBookmarksView &)m_view;
-    bookmarksFrame.ParseBookmarks(attribute);
+    CBookmarksView &bookmarksView = (CBookmarksView &)m_view;
```
[the remark this answers is not in the document]
Comment (dehilsterlexis) [reply eclide/BookmarksFrame.cpp]: This truly wins the stupid award. Oh my goodness. Yes, you are right. Probably had a reason why I was casting way back but probably not.

File: eclide/BookmarksFrame.cpp [hunk 4]
```diff
@@ -37,11 +37,92 @@ bool CBookmarksFrame::OnContext(CMenu &/*menu*/, CPoint &/*pt*/)
 }
 void CBookmarksFrame::ParseBookmarksEcl(std::_tstring ecl, std::_tstring user, std::_tstring id, std::_tstring inModule, std::_tstring inAttributeName, IAttributeType *attrType)
 {
-    CBookmarksView &bookmarksFrame = (CBookmarksView &)m_view;
-    bookmarksFrame.ParseBookmarksEcl(ecl, user, id, inModule, inAttributeName, attrType);
+    CBookmarksView &bookmarksView = (CBookmarksView &)m_view;
```
[the remark this answers is not in the document]
Comment (dehilsterlexis) [reply eclide/BookmarksFrame.cpp]: Yep. Totally right.

File: comms/EclCC.cpp [hunk 5]
```diff
@@ -410,6 +409,31 @@ class CEclCC : public IEclCC, public clib::CLockableUnknown
...
+            {
+                folderPath = p2.parent_path().c_str();
+                module = pathStr.substr(folderPath.size() + 1, pathStr.size() - folderPath.size() - attr.size() - 2);
+                boost::algorithm::replace_all(module, _T("\\"), _T("."));
+                break;
+            }
+        }
+        attrName = pathToWString(p.stem());
+        return module.c_str();
+    }
```
[the remark this answers is not in the document]
Comment (dehilsterlexis) [reply comms/EclCC.cpp]: Will look into it. I did think of something like that but opted not to change but will revisit.

File: comms/EclCC.cpp [hunk 6]
```diff
@@ -410,6 +409,31 @@ class CEclCC : public IEclCC, public clib::CLockableUnknown
 #endif
     }
 
+    const TCHAR * ModuleFromPath(const std::_tstring & pathStr, std::_tstring & attrName, std::_tstring & module)
```
[the remark this answers is not in the document]
Comment (dehilsterlexis) [reply comms/EclCC.cpp]: Actually, I need to rename some variables to retAttrName, retModuleName and rename the function.

**Comment by richardkchapman** [comment]
@GordonSmith

File: comms/ModuleHelper.cpp [hunk 7]
```diff
@@ -108,3 +115,88 @@ unsigned int CModuleHelper::GetQualifiedLabel(StdStringVector & results) const
...
+    return retLabel.c_str();
+}
+
+const CComPtr<IAttribute> CModuleHelper::GetAttribute(const std::_tstring & module, const std::_tstring & attribute, const std::_tstring & attrExt, std::_tstring & path) const
+{
+    CComPtr<IRepository> rep = AttachRepository();
+    CComPtr<IAttribute> attr = rep->GetAttributeFast(module.c_str(), attribute.c_str(), AttributeTypeFromExtension(attrExt));
+    if (attr == NULL)
+    {
+        //IAttribute * attribute = CreateDiskAttribute(rep, module, attribute, attrExt, path);
```
Comment (GordonSmith) [inline comms/ModuleHelper.cpp]: commented code

File: eclide/BookmarksFrame.cpp [hunk 8]
```diff
@@ -37,11 +37,92 @@ bool CBookmarksFrame::OnContext(CMenu &/*menu*/, CPoint &/*pt*/)
...
+    bookmarksView.ParseBookmarksEcl(ecl, user, id, inModule, inAttributeName, attrType);
 }
 void CBookmarksFrame::ParseBookmarksEcl(IAttribute *attribute)
 {
-    CBookmarksView &bookmarksFrame = (CBookmarksView &)m_view;
-    bookmarksFrame.ParseBookmarks(attribute);
+    CBookmarksView &bookmarksView = (CBookmarksView &)m_view;
+    bookmarksView.ParseBookmarks(attribute);
+}
+void CBookmarksFrame::ParsePathBookmarksEcl(std::_tstring pathStr)
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply eclide/BookmarksFrame.cpp]: @dehilsterlexis - bump

File: eclide/BookmarksView.cpp [hunk 9]
```diff
@@ -807,7 +854,7 @@ void CBookmarksView::OnLoadFile(bool mergeFlag)
             }
         }
 
-        DoRefresh(NULL, 0);
+        //DoRefresh(NULL, 0);
```
Comment (GordonSmith) [inline eclide/BookmarksView.cpp]: Commented code.

File: eclide/BookmarksView.h [hunk 10]
```diff
@@ -94,12 +94,19 @@ class CBookmarksView:
     std::_tstring FindTag(std::_tstring line, std::_tstring tag, int &index, bool zeroIndex = false);
     void SetMarks(bool val);
     void DeleteMarkedBookmarks(bool val);
-    void SetMarks(std::_tstring inModule, std::_tstring inAttributeName, bool val);
-    void DeleteMarkedBookmarks(std::_tstring inModule, std::_tstring inAttributeName, bool val);
+    void SetMarks(std::_tstring inPath, std::_tstring inModule, std::_tstring inAttributeName, bool val);
+    void DeleteMarkedBookmarks(std::_tstring inPath, std::_tstring inModule, std::_tstring inAttributeName, bool val);
```
Comment (GordonSmith) [inline eclide/BookmarksView.h]: should be `const std::_tstring &`?

File: eclide/mainfrm.cpp [hunk 11]
```diff
@@ -2188,16 +2351,6 @@ bool CMainFrame::UIUpdateMenuItems()
...
 
-BOOL CMainFrame::DoFileOpen(const CString & sPathName, int row, int col, int len)
-{
-    boost::filesystem::path path = stringToPath(static_cast<const TCHAR *>(sPathName));
-    std::_tstring filename = pathToWString(path.leaf());
-    CComPtr<IRepository> rep = AttachRepository();
-    AttrInfo attrInfo;
-    attrInfo.AttributeType = ExtensionWithoutDot(filename);
-    return OpenFileBuilderMDI(this, attrInfo, sPathName, rep->CreateIWorkspaceItem(WORKSPACE_ITEM_BUILDER, filename, static_cast<const TCHAR *>(sPathName)), false, row, col, len);
-}
```
Comment (GordonSmith) [inline eclide/mainfrm.cpp]: Is this not needed for opening an editor at a specific location (syntax errors)?

**Comment by dehilsterlexis** [comment]
@GordonSmith I tested an installed version for local, github repository, and remote repository with mixed local and remote files. All working. Ready for review.

**Comment by GordonSmith** [comment]
@dehilsterlexis can you rebase?

**Comment by dehilsterlexis** [comment]
@GordonSmith rebased

File: eclide/BookmarksView.cpp [hunk 12]
```diff
@@ -497,64 +490,97 @@ void CBookmarksView::ParseBookmarksEcl(const std::_tstring & ecl, const std::_ts
...
+{
+    if (boost::filesystem::exists(pathStr))
+    {
+        std::_tstring label, module, attr, attrType;
+        attr = attrName;
+        CModuleHelper modHelper(_T(""));
+        modHelper.ModuleAttrFromPath(pathStr, label, module, attr, attrType);
+        return module;
+    }
+    return NULL;
```
Comment (GordonSmith) [inline eclide/BookmarksView.cpp:583]: Return type is wrong.

File: eclide/BookmarksView.h (hunk 10 above)
Comment (GordonSmith) [reply eclide/BookmarksView.h:97]: @dehilsterlexis bump

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

**Pull request merged**